### ISSUE TITLE
Pixel Recurrent Neural Network

### ISSUE THREAD

**Issue by omoindrot** [omoindrot/papers#4]
http://arxiv.org/abs/1601.06759
DeepMind
## Objective

**Unsupervised learning**
Predict the rest of an image from a part of it (generative modeling).
They use a two-dimensional LSTM to capture the sense of the image.
## Model

Like in RNN when trying to predict the next word, we first have to build a **representation of the likelihood** of each image.  
We model the probability for an image to exist as:
![proba](https://cloud.githubusercontent.com/assets/13662086/14916220/77eaf83c-0e18-11e6-9e34-74127726b1af.png)
where pixels are taken row by row.

For each pixel, we predict the RGB colors one by one: 
![proba pixel](https://cloud.githubusercontent.com/assets/13662086/14916243/9fd29d50-0e18-11e6-9d68-104b3ad137fd.png)

and for each channel, we output a 256-d vector of probability with a final softmax.
### Row LSTM

The Row LSTM is an LSTM iterating on the rows.
The matrix input-to-state is a 1-dim convolutional layer with kernel 3.
The matrix state-to-state is a convolutional layer with kernel 3. Therefore, the receptive field has a triangular shape:
![receptive field](https://cloud.githubusercontent.com/assets/13662086/14916549/94dcbd84-0e1a-11e6-88ac-3a8fc678f285.png)
### Diagonal bi-LSTM

The LSTM iterates on the rows.
The matrix input-to-state is a convolutional layer with kernel 1.
The matrix state-to-state of each component of the bi-LSTM is a convolutional layer with kernel 2, taking the previous states on the left and on the top. Therefore, the receptive field has all the previous pixels:
![receptive field 2](https://cloud.githubusercontent.com/assets/13662086/14916819/154dcb10-0e1c-11e6-92cc-5ef92cbac2d1.png)
### Residual Connections

They use multiple LSTMs on top of each other (12 in total), so to make the model trainable they add residual connections like in ResNets.
![residual connections](https://cloud.githubusercontent.com/assets/13662086/14916992/18eaa7e2-0e1d-11e6-8bbe-85d1c7b1d683.png)
### Masked Convolutions

When seeing the input image, we want to predict color R of the pixel knowing only the previous pixels. We then want to predict color G given previous pixels and color R of the pixel. Therefore we apply a mask 0 to zero the weights on the unwanted input.
In the network then, the color R can only be predicted given previous R (previous G or B contains the input R and G data !).
![mask](https://cloud.githubusercontent.com/assets/13662086/14917376/097d28a0-0e1f-11e6-8971-1ff540e3a30a.png)
## Results

They improved state of the art on MNIST and CIFAR-10, and created a new benchmark for ImageNet. The loss is computed as negative log likelihood on the pixel values.
Their best result is `3.0`, which corresponds to predicting 0.05 on the good pixel with the last softmax.
